### PR TITLE
WH: Remove slave dependency on warehouseDB

### PR DESCRIPTION
**Pull request by psrikanth88**
**Fixes** # (*issue*)

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
## Checklist:
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: warehouse/slave.go [hunk 1]
```diff
@@ -559,13 +560,12 @@ func processClaimedJob(claimedJob pgnotifier.ClaimT) {
 		return
 	}
 	job.BatchID = claimedJob.BatchID
-	ids, err := processStagingFile(job)
-	job.setStagingFileStatus(err)
```
Comment (sivashanmukh) [inline warehouse/slave.go:563]: Why is this removed? This was added in a previous PR, right?

Comment (sivashanmukh) [inline warehouse/slave.go:563]: If this function is not used anywhere else, we should remove the function altogether.

Comment (psrikanth88) [reply warehouse/slave.go:563]: removed

File: warehouse/upload.go [hunk 2]
```diff
@@ -1153,6 +1184,58 @@ func (job *UploadJobT) createLoadFiles(generateAll bool) (startLoadFileID int64,
 	return startLoadFileID, endLoadFileID, nil
 }
 
+func (job *UploadJobT) setStagingFileSuccess(stagingFileIDs []int64) {
+	sqlStatement := fmt.Sprintf(`UPDATE %s SET status=$1, updated_at=$2 WHERE id=ANY($3)`, warehouseutils.WarehouseStagingFilesTable)
```
Comment (SumanthPuram) [inline warehouse/upload.go:1188]: Minor comment, ignore if not needed. Should we use `where id IN $3` instead of `where id=ANY($3)`

Both should work, but `IN` seems to be more intuitive and we use that across other modules.

File: warehouse/upload.go [hunk 3]
```diff
@@ -1131,19 +1133,48 @@ func (job *UploadJobT) createLoadFiles(generateAll bool) (startLoadFileID int64,
...
+				if len(output) == 0 {
+					pkgLogger.Errorf("[WH]: No LoadFiles returned by wh worker")
+					continue
+				}
+				loadFiles = append(loadFiles, output...)
+				successfulStagingFileIDs = append(successfulStagingFileIDs, resp.JobID)
+			}
+			err = job.bulkInstertLoadFileRecords(loadFiles)
+			if err != nil {
+				saveLoadFileErrs = append(saveLoadFileErrs, err)
```
[the remark this answers is not in the document]
Comment (sivashanmukh) [reply warehouse/upload.go:1159]: If we have a continue, wg.Done() should be done before that. What's the purpose of continue? Do you intend that we skip job.setStagingFileSuccess?

File: warehouse/upload.go (hunk 2 above)
Comment (psrikanth88) [reply warehouse/upload.go:1188]: ANY seems to be more verstaile https://stackoverflow.com/a/34627688
Also pq.Array can be used instead of us preparing a comma separated list of values

File: warehouse/upload.go [hunk 4]
```diff
@@ -1131,19 +1133,48 @@ func (job *UploadJobT) createLoadFiles(generateAll bool) (startLoadFileID int64,
...
 			}
+			job.setStagingFileSuccess(successfulStagingFileIDs)
 			wg.Done()
 		})
 	}
 
 	wg.Wait()
 
+	if len(saveLoadFileErrs) > 0 {
```
[the remark this answers is not in the document]
Comment (psrikanth88) [reply warehouse/upload.go:1168]: will take this up in different PR

File: warehouse/upload.go (hunk 2 above)
Comment (sivashanmukh) [reply warehouse/upload.go:1188]: I would suggest that we start using ANY here and switch everywhere else with IN. It seems to be a better way to construct queries from code.

**Review by sivashanmukh — APPROVED** [review]
LGTM

File: warehouse/upload.go (hunk 2 above)
Comment (SumanthPuram) [reply warehouse/upload.go:1188]: https://stackoverflow.com/a/34627688 ->

This says IN uses indexes and ANY doesn't.

**Review by SumanthPuram — APPROVED** [review]

File: warehouse/upload.go (hunk 2 above)
Comment (sivashanmukh) [reply warehouse/upload.go:1188]: I too was confused when I saw that and had done an EXPLAIN on both variants in our query like so
`EXPLAIN ANALYZE UPDATE wh_staging_files SET status='success', updated_at='2021-01-01' WHERE id=ANY('{3,2,1}');
`
`EXPLAIN ANALYZE UPDATE wh_staging_files SET status='success', updated_at='2021-01-01' WHERE id IN (3,2,1);`

Both of them had given the same query plan. I had then read the reply again and what you mentioned along with the other link below it are special cases where the index wasn't used and we don't really use IN or ANY in that way afaik.

The one that you saw, points to this link -> https://dba.stackexchange.com/questions/125413/index-not-used-with-any-but-used-with-in/125500#125500 which is the special case the answer discusses.

Comment (SumanthPuram) [reply warehouse/upload.go:1188]: Just checked this. 

For composite indexes, ANY is not using the index but IN is using it. Since this is a primary key, we will go ahead with ANY.

**Review by SumanthPuram — APPROVED** [review]
